src/nlp_expr.jl

```diff
@@ -0,0 +1,979 @@
+end
+
+function Base.convert(
+    ::Type{<:GenericNonlinearExpr},
+    x::GenericQuadExpr{C,V},
+) where {C,V}
+    args = Any[]
+    for (variable, coef) in x.aff.terms
+        if isone(coef)
+            push!(args, variable)
+        elseif !iszero(coef)
+            push!(args, GenericNonlinearExpr{V}(:*, coef, variable))
+        end
+    end
+    for (pair, coef) in x.terms
+        if isone(coef)
+            push!(args, GenericNonlinearExpr{V}(:*, pair.a, pair.b))
+        elseif !iszero(coef)
+            push!(args, GenericNonlinearExpr{V}(:*, coef, pair.a, pair.b))
+        end
+    end
+    if !iszero(x.aff.constant) || isempty(args)
+        push!(args, x.aff.constant)
+    end
+    if length(args) == 1
+        return args[1]
+    end
+    return GenericNonlinearExpr{V}(:+, args)
+end
+
+function _MA.promote_operation(
+    ::Union{typeof(+),typeof(-),typeof(*)},
+    ::Type{GenericNonlinearExpr{V}},
+    ::Type{<:AbstractJuMPScalar},
+) where {V<:AbstractVariableRef}
+    return GenericNonlinearExpr{V}
+end
+
+function _MA.promote_operation(
+    ::Union{typeof(+),typeof(-),typeof(*)},
+    ::Type{<:AbstractJuMPScalar},
+    ::Type{GenericNonlinearExpr{V}},
+) where {V<:AbstractVariableRef}
+    return GenericNonlinearExpr{V}
+end
+
+function _MA.promote_operation(
+    ::Union{typeof(+),typeof(-),typeof(*)},
+    ::Type{GenericNonlinearExpr{V}},
+    ::Type{GenericNonlinearExpr{V}},
+) where {V<:AbstractVariableRef}
+    return GenericNonlinearExpr{V}
+end
+
+function _MA.promote_operation(
+    ::Union{typeof(+),typeof(-),typeof(*)},
+    ::Type{GenericNonlinearExpr{U}},
+    ::Type{GenericNonlinearExpr{V}},
+) where {U<:AbstractVariableRef,V<:AbstractVariableRef}
+    return error(
+        "Unable to promote two different types of nonlinear expression",
+    )
+end
+
+"""
+    UserDefinedFunction(head::Symbol, func::Function)
+
+A struct representing a user-defined function named `head`.
+
+This function must have already been added to the model using
+[`add_user_defined_function`](@ref) or [`@register`](@ref).
+
+## Example
+
+```jldoctest
+julia> model = Model();
+
+julia> @variable(model, x)
+x
+
+julia> f(x::Float64) = x^2
+f (generic function with 1 method)
+
+julia> ∇f(x::Float64) = 2 * x
+∇f (generic function with 1 method)
+
+julia> ∇²f(x::Float64) = 2.0
+∇²f (generic function with 1 method)
+
+julia> @register(model, foo, 1, f, ∇f, ∇²f)
+UserDefinedFunction{typeof(f)}(:foo, f)
+
+julia> bar = UserDefinedFunction(:foo, f)
+UserDefinedFunction{typeof(f)}(:foo, f)
+
+julia> @objective(model, Min, bar(x))
+foo(x)
+
+julia> bar(2.0)
+4.0
+```
+"""
+struct UserDefinedFunction{F}
+    head::Symbol
+    func::F
+end
+
+function (f::UserDefinedFunction)(args...)
+    if any(Base.Fix2(isa, AbstractJuMPScalar), args)
+        return GenericNonlinearExpr(f.head, Any[a for a in args])
+    end
+    return f.func(args...)
+end
+
+"""
+    add_user_defined_function(
+        model::Model,
+        op::Symbol,
```

Comment:
I would consider having `; name = Symbol(f)` at the end.
I know this allows the user doing `add_user_defined_function(model, 1, x -> x^2)` which would have a bad name but it's the same for anonymous variables when the user uses `JuMP.add_variable` without the `name` keyword argument.
This name keyword would also be consistent with `add_variable` and `add_constraint` macros.

Reply:
We don't need convenience. In this case, we need clarity, and the user can always use `add_user_defined_fuction(model, gensym(), ...)` if they don't care about the `Symbol`.

We don't need this to match the macros.

And no one has ever asked this for the current `register` function in JuMP, which also required something like `register(model, :f, 1, f; autodiff = true)`.

Reply:
I don't know why we wouldn't need the convenience. We are making a macro just to help the users not have to worry about name clash with the function name in global scope. Users that want to use the same name, or users who feel that the readability loss of using a macro is not worth it will use the `register` function. Having to add this symbol everytime is quite cumbersome and error prone.

> And no one has ever asked this for the current `register` function in JuMP, which also required something like `register(model, :f, 1, f; autodiff = true)`.

I must admit that this always left me confused why you had to specify as symbol and it was not taking the function name by default.

Reply:
For what it is worth, I would agree with @blegat. It would be nice to make the name an optional argument where the default is to use `Symbol(f)`. This is what InfiniteOpt and Symbolics currently do.

Reply:
Okay. I'll have a go

Reply:
Done